src/ray/id.h

```diff
@@ -2,44 +2,128 @@
 #define RAY_ID_H_
 
 #include <inttypes.h>
+#include <limits.h>
 
+#include <chrono>
 #include <cstring>
+#include <mutex>
+#include <random>
 #include <string>
 
 #include "plasma/common.h"
 #include "ray/constants.h"
+#include "ray/util/logging.h"
 #include "ray/util/visibility.h"
 
 namespace ray {
 
-class RAY_EXPORT UniqueID {
+class DriverID;
+class UniqueID;
+
+// Declaration.
+std::mt19937 RandomlySeededMersenneTwister();
+uint64_t MurmurHash64A(const void *key, int len, unsigned int seed);
+
+// Change the compiler alignment to 1 byte (default is 8).
+#pragma pack(push, 1)
+
+template <typename T>
+class BaseID {
  public:
-  UniqueID();
-  UniqueID(const plasma::UniqueID &from);
-  static UniqueID from_random();
-  static UniqueID from_binary(const std::string &binary);
-  static const UniqueID &nil();
+  BaseID();
+  static T from_random();
+  static T from_binary(const std::string &binary);
```

Comment:
Just realized that we are not using `CamelCase` convention for method names here..
We can fix this later

Reply:
Yes, Java code has already used `CamelCase`. This PR is too big to do that. I will do that in next PR.